### PR TITLE
BigQuery updates

### PR DESCRIPTION
**Pull request by quartzmo**
This PR will merge the `bigquery` topic branch (with commits dating from the beginning of August) into `master`.

The PR is functionally complete. The only outstanding update planned at this point is for improving some of the older documentation (#1691).

Commits of particular interest are:

* [Add QueryJob#data](https://github.com/GoogleCloudPlatform/google-cloud-ruby/commit/959d0452fdfac3d0c7b564ecdf8f92f461a3b49b) (wait for job completion) - Issue #1623, PR #1657
* [BigQuery custom error retry and backoff logic](https://github.com/GoogleCloudPlatform/google-cloud-ruby/commit/3d888933f8a67c2b17402b2bae8c6943bdcc4ae4) - Issue #1715, PR #1732
* [Update Job#wait_until_done!](https://github.com/GoogleCloudPlatform/google-cloud-ruby/commit/7e2db72492ab308ba8675fe918838473d6ed9854) (wait for job completion) - Issue #1626, #1720, PR #1751
* [Asynchronous Insert](https://github.com/GoogleCloudPlatform/google-cloud-ruby/commit/28be92776fb5f41b5588b9fe19370f69100e5ce9) (batching for streaming data) - Issue #1736, PR #1744

/cc @jjolma

### CODE REVIEW

**Comment by blowmage** [comment]
These are my commits.

**Comment by blowmage** [comment]
@jjolma We cannot add you as a reviewer because of some GitHub permission, but we were told that you should review these changes before they are released. Can you please take a look and add a review?

**Comment by blowmage** [comment]
@jjolma We have been instructed to ask for your review of this PR. Is this something you can review?

We will plan on merging and release tomorrow if we don't hear from you.

**Comment by jjolma** [comment]
@blowmage sorry I missed this. Yes, I should be able to review it in the morning if that's okay.

**Comment by blowmage** [comment]
Sounds great!

File: google-cloud-bigquery/lib/google/cloud/bigquery/service.rb [hunk 1]
```diff
@@ -573,11 +650,84 @@ def mime_type_for file
...
+              return true if retry_error_reason? result
+            end
+            false
+          end
+
+          def retry_error_reason? err_body
+            err_hash = JSON.parse err_body
+            Array(err_hash["error"]["errors"]).each do |error|
+              return true if @reasons.include? error["reason"]
+            end
```
Comment (jjolma) [inline google-cloud-bigquery/lib/google/cloud/bigquery/service.rb]: The logic should be: retry if *all* errors are retryable, not if *any* are.

**Comment by quartzmo** [comment]
@jjolma Thank you for the review! We really appreciate it. These three points seem clear and we will address them in updates to this PR.

**Comment by blowmage** [comment]
> Also, what do we need to do to fix the permissions probably that prevents adding me as a reviewer (that @blowmage mentioned earlier)?

Honestly, I don't know the answer to that yet. I was hoping that since you added a review to this PR I could add you to others, but I just tried with #1766 and could not add you.

**Comment by blowmage** [comment]
@jjolma We have addressed the issues you pointed out. Can you update your review of this PR?

**Comment by jjolma** [comment]
LGTM. Thanks!

**Pull request merged**